src/navigation/RootStackNavigation.tsx

```diff
@@ -84,104 +98,193 @@ const RootStackNavigation = () => {
       </Host>
     </GestureHandlerRootView>
   )
 }
 
 export default RootStackNavigation
+
+const AppUnlockHandler = ({ children }: { children: ReactNode }) => {
+  const dispatch = useAppDispatch()
+  const appState = useRef(AppState.currentState)
+  const lastNavigationState = useAppSelector((s) => s.app.lastNavigationState)
+  const isCameraOpen = useAppSelector((s) => s.app.isCameraOpen)
+  const walletMnemonic = useAppSelector((s) => s.wallet.mnemonic)
+  const addressesStatus = useAppSelector((s) => s.addresses.status)
+  const navigation = useNavigation<NavigationProp<RootStackParamList>>()
+
+  const [isAppStateChangeCallbackRegistered, setIsAppStateChangeCallbackRegistered] = useState(false)
+
+  const unlockApp = useCallback(async () => {
+    if (walletMnemonic) return
+
+    try {
+      const deviceHasBiometricsData = await isEnrolledAsync()
+      const walletMetadata = await getWalletMetadata()
+      let isBioEnabled = await loadBiometricsSettings()
+
+      // Disable biometrics if needed
+      if (isBioEnabled && !deviceHasBiometricsData) {
+        await disableBiometrics()
+        await storeBiometricsSettings(false)
+        dispatch(biometricsToggled(false))
+        isBioEnabled = false
+      }
+
+      const wallet = await getStoredWallet()
+
+      if (!wallet) {
+        if (lastNavigationState) {
+          setNavigationState(lastNavigationState)
+        } else {
+          navigation.navigate('LandingScreen')
+        }
+      } else {
+        if (isBioEnabled) {
+          const addressesToInitialize =
+            addressesStatus === 'uninitialized' ? await deriveWalletStoredAddresses(wallet) : []
+          dispatch(walletUnlocked({ wallet, addressesToInitialize, contacts: walletMetadata?.contacts ?? [] }))
+
+          lastNavigationState ? setNavigationState(lastNavigationState) : resetNavigationState()
+        } else {
+          navigation.navigate('LoginWithPinScreen')
+        }
+      }
+
+      // TODO: Revisit error handling with proper error codes
+    } catch (e: unknown) {
+      const error = e as { message?: string }
+
+      if (error.message?.includes('User canceled')) {
+        Alert.alert('Authentication required', 'Please authenticate to unlock your wallet.', [
+          { text: 'Try again', onPress: unlockApp }
+        ])
+      } else {
+        console.error(e)
+      }
+    }
+  }, [addressesStatus, dispatch, lastNavigationState, navigation, walletMnemonic])
+
+  useEffect(() => {
+    const handleAppStateChange = (nextAppState: AppStateStatus) => {
+      if (nextAppState === 'background' && walletMnemonic && !isCameraOpen) {
+        navigation.navigate('SplashScreen')
+        dispatch(appBecameInactive())
+      } else if (nextAppState === 'active' && !walletMnemonic && !isCameraOpen) {
+        unlockApp()
+      }
+
+      appState.current = nextAppState
+    }
+
+    if (!isAppStateChangeCallbackRegistered) {
+      handleAppStateChange(appState.current)
+    }
```

Comment:
When the app is initially launched (either after a fresh install, a device restart, or if the user quits the app and launches it again), the `'change'` event of `AppState` does not get triggered (kinda makes sense). Thus, we manually need to run the code that handles the event, only for the first time.